lib/src/java/facter.cc

```diff
@@ -0,0 +1,175 @@
+            return true;
+        });
+        return hashmap;
+    }
+    return nullptr;
+}
+
+extern "C" {
+    LIBFACTER_EXPORT jint JNI_OnLoad(JavaVM* vm, void* reserved)
+    {
+        JNIEnv* env;
+        if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+            return JNI_ERR;
+        }
+        // For the classes, we need global refs
+        object_class = find_class(env, "java/lang/Object");
+        if (!object_class) {
+            return JNI_ERR;
+        }
+        long_class = find_class(env, "java/lang/Long");
+        if (!long_class) {
+            return JNI_ERR;
+        }
+        double_class = find_class(env, "java/lang/Double");
+        if (!double_class) {
+            return JNI_ERR;
+        }
+        boolean_class = find_class(env, "java/lang/Boolean");
+        if (!boolean_class) {
+            return JNI_ERR;
+        }
+        hash_class = find_class(env, "java/util/HashMap");
+        if (!hash_class) {
+            return JNI_ERR;
+        }
+
+        // For the method ids, we can keep these cached as is
+        long_constructor = env->GetMethodID(long_class, "<init>", "(J)V");
+        double_constructor = env->GetMethodID(double_class, "<init>", "(D)V");
+        boolean_constructor = env->GetMethodID(boolean_class, "<init>", "(Z)V");;
+        hash_constructor = env->GetMethodID(hash_class, "<init>", "(I)V");
+        hash_put = env->GetMethodID(hash_class, "put", "(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;");
+
+        setup_logging(boost::nowide::cerr);
+        set_level(level::warning);
+
+        facts_collection.reset(new collection());
+        facts_collection->add_default_facts();
+        facts_collection->add_external_facts();
+        return JNI_VERSION_1_6;
+    }
+
+    LIBFACTER_EXPORT void JNI_OnUnload(JavaVM* vm, void* reserved)
+    {
+        // Delete the fact collection
+        facts_collection.reset();
+
+        JNIEnv* env;
+        if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+            return;
+        }
+        // Free all of the global references above
+        if (object_class) {
+            env->DeleteGlobalRef(object_class);
+            object_class = nullptr;
+        }
+        if (long_class) {
+            env->DeleteGlobalRef(long_class);
+            long_class = nullptr;
+        }
+        if (double_class) {
+            env->DeleteGlobalRef(double_class);
+            double_class = nullptr;
+        }
+        if (boolean_class) {
+            env->DeleteGlobalRef(boolean_class);
+            boolean_class = nullptr;
+        }
+        if (hash_class) {
+            env->DeleteGlobalRef(hash_class);
+            hash_class = nullptr;
+        }
+    }
+
+    LIBFACTER_EXPORT jobject JNICALL Java_com_puppetlabs_Facter_lookup(JNIEnv* env, jclass klass, jstring name)
+    {
+        // Ensure initialized
+        if (!facts_collection) {
```

Comment:
Probably not; it's unlikely we'll get called on this export without being successfully loaded through JNI, but it doesn't really hurt.